Add peak labels to free energy plots

SCiarella opened the pull request:
This branch covers #203.
I have used a KD-tree to precompute a dictionary that maps each site to the closest peak.
Then after finding the path, I implemented a method to use the dictionary to identify which peaks get explored.
This information gets automatically plotted on the x-axis of the energy landscape:
![fig1](https://github.com/GEMDAT-repos/GEMDAT/assets/58949181/11914d94-e2d4-4d6b-a4f8-5d749d9b1cff)
![fig2](https://github.com/GEMDAT-repos/GEMDAT/assets/58949181/5c126899-eec8-4cfd-8c2a-48e70570223e)

stefsmeets commented:
To my understanding the goal for the related issue is to link the coordinates back to the site labels from the cif file.

i.e. `48h` instead of `(28, 15, 10)`

SCiarella commented:
Now the path is compared to the real structure obtained from `load_known_material(...)` instead of the peaks of the density.
The figure have been updated in order to plot the site label on the top, and its Cartesian coordinates on the bottom:

![1](https://github.com/GEMDAT-repos/GEMDAT/assets/58949181/b85674e7-ab6a-493f-b91d-b9c4db991b29)
![2](https://github.com/GEMDAT-repos/GEMDAT/assets/58949181/5daa2006-e29e-44ff-81fc-8b8f64e43b3f)

stefsmeets reviewed (commented):
Hi @SCiarella , thanks for this PR. I just had a look and left some feedback. The plots look nice, though I'm wondering if it makes sense that the diffusing atom passes through so many cages so quickly.

My overall feeling is that this PR is not complete yet, and I'm not quite sure how this is supposed to be used.

- Can you add some tests for the plot (see `tests/integration/plot_test.py`)?
- Can you add some instructions for how this code is supposed to be used?
- The kdtree looks for the nearest atom for every voxel, shouldn't there be based some sort of distance threshold?

SCiarella commented on src/gemdat/path.py:
```diff
@@ -124,7 +171,7 @@ def optimal_path(
     path_energy = [F_graph.nodes[node]['energy'] for node in optimal_path]
 
-    path = Pathway(optimal_path, path_energy)
+    path = Pathway(optimal_path, [], path_energy, [], [])
```
Thanks for the suggestion. Now Pathway uses optional arguments that are empty by defauls

SCiarella commented on src/gemdat/path.py:
```diff
@@ -34,6 +58,29 @@ def wrap(self, F: np.ndarray):
+            raise ValueError(
+                'Cartesian coordinates of the path sites are not defined')
+
+        self.nearest_structure_label = [
+            structure.labels[structure_map.get(site, None)]
+            for site in self.cart_sites
+        ]
+        self.nearest_structure_coord = [
+            structure.cart_coords[structure_map.get(site, None)]
+            for site in self.cart_sites
+        ]
+
```
I have refactored these operation to use a single loop, so I would keep them together.
The `path_over_structure()` method is required to plot the labels in `energy_along_path()`, but since I wanted to leave the label optiona, this is not explicitely called.

SCiarella commented on src/gemdat/path.py:
```diff
@@ -15,13 +18,34 @@ class Pathway:
-    path: list[tuple]
-        List of coordinates of the sites definint the path
-    path_energy: list[float]
+    sites: list[tuple]
+        List of voxel coordinates of the sites defining the path
+    cart_sites: list[tuple]
+        List of cartesian coordinates of the sites defining the path
+    energy: list[float]
         List of the energy along the path
+    nearest_structure_coord: list[np.ndarray]
+        List of cartesian coordinates of the closest site of the reference structure
+    nearest_structure_label: list[str]
+        List of the label of the closest site of the reference structure
```
Thanks for the suggestion! the attributes are now optional

SCiarella commented on src/gemdat/path.py:
```diff
@@ -15,13 +18,34 @@ class Pathway:
+        vol : Volume
+            Volume object containing the grid information
+        """
+        self.cart_sites = [
```
I have changed this method to use Lattice.get_cartesian_coords(). This should solve the issue of non orthorombic lattices

SCiarella commented on src/gemdat/plots/matplotlib/_paths.py:
```diff
@@ -18,10 +18,37 @@ def energy_along_path(*, path: Pathway) -> plt.Figure:
+    ax.set(ylabel='Free energy [eV]')
+    if path.nearest_structure_label:
+        # Remove repeated labels that would bloat the figure
+        clean_xlabel = [
+            f'{", ".join([f"{coord:.1f}" for coord in path.nearest_structure_coord[i]])}'
+            if (path.nearest_structure_coord[i]
+                != path.nearest_structure_coord[i - 1]).any() else ''
+            for i in range(len(path.sites))
+        ]
```
Sure! The piece of code is now an explicit loop, much easier to read.

SCiarella commented on src/gemdat/plots/matplotlib/_paths.py:
```diff
@@ -18,10 +18,37 @@ def energy_along_path(*, path: Pathway) -> plt.Figure:
+        ax.set_xticklabels([clean_xlabel[i] for i in non_empty_ticks],
+                           rotation=45)
+        # and add on top the site labels
+        clean_xlabel_up = [
+            f'{path.nearest_structure_label[i]}' if
+            (path.nearest_structure_coord[i]
+             != path.nearest_structure_coord[i - 1]).any() else ''
+            for i in range(len(path.sites))
+        ]
```
The list is now populated in an explicit loop

SCiarella commented on src/gemdat/volume.py:
```diff
@@ -122,6 +123,51 @@ def find_peaks(
+                            dy, self.lattice.c * dz
+                        ]) * self.resolution)
+                    # store the id of the site in the original structure
+                    periodic_ids.extend(
+                        [i for i in range(len(structure.cart_coords))])
```
Thanks, it works!

SCiarella commented on src/gemdat/volume.py:
```diff
@@ -122,6 +123,51 @@ def find_peaks(
+        # In order to accomodate the periodicity, include the images of the structure sites
+        periodic_structure = []
+        periodic_ids = []
+        for dx in [-1, 0, 1]:
+            for dy in [-1, 0, 1]:
+                for dz in [-1, 0, 1]:
```
Yes, it is indeed an optimal way to loop. I have implemented this suggestion.

SCiarella commented on src/gemdat/volume.py:
```diff
@@ -122,6 +123,51 @@ def find_peaks(
 
         return coords[:, 0:3].astype(int)
 
+    def nearest_structure_reference(self, structure: Structure):
```
I have refactored this method and now it is not returning anything

SCiarella commented on src/gemdat/volume.py:
```diff
@@ -122,6 +123,51 @@ def find_peaks(
+            for dy in [-1, 0, 1]:
+                for dz in [-1, 0, 1]:
+                    periodic_structure.extend(
+                        structure.cart_coords + np.array([
+                            self.lattice.a * dx, self.lattice.b *
+                            dy, self.lattice.c * dz
+                        ]) * self.resolution)
```
I have used `get_fractional_coords`, but only after the summation, because the summation is easier in cartesian coordinates

stefsmeets replied:
Yeah, my bad, I meant `.get_cartesian_coords()` 😅

SCiarella commented:
As discussed during the meeting on 09/01/24, I have implemented a color-coding to easily distinguish the closest site to the path:
![a](https://github.com/GEMDAT-repos/GEMDAT/assets/58949181/1cab6f34-c3b7-4419-8782-ca5b05c58548)

In this last series of commits, I have also added a test for this plotting function.

stefsmeets commented:
I'll review this next week when I'm back!

stefsmeets commented on src/gemdat/volume.py:
```diff
@@ -122,6 +129,36 @@ def find_peaks(
+        periodic_ids: list[int] = []
+        images = np.mgrid[-1:2, -1:2, -1:2].reshape(3, -1).T
+        for dx, dy, dz in images:
+            periodic_structure.extend(
+                self.lattice.get_fractional_coords(
+                    structure.cart_coords + np.array([
+                        self.lattice.a * dx, self.lattice.b *
+                        dy, self.lattice.c * dz
+                    ]) * self.resolution))
```
I can't understand what this is trying to do. It's going from fractional to cartesian and back to fractional... Is this just adding the translations? 

Why not something like this:

```suggestion
            periodic_structure.extend(structure.frac_coords + np.array([dx, dy, dz])
```

stefsmeets commented on src/gemdat/volume.py:
```diff
@@ -122,6 +129,36 @@ def find_peaks(
+        # Create a KD-tree from the structure
+        kd_tree = cKDTree(periodic_structure)
+
+        self.nearest_structure_tree = kd_tree
+        self.nearest_structure_map = periodic_ids
```
It's not obvious to me that these are stored on `Volume`. Might make more sense to return these.

SCiarella replied:
Same ad the coordinates, they have been transformed to functions

stefsmeets reviewed (changes requested):
I don't understand what is going on in this PR, it also includes the changes from #239, which makes this difficult to review. I tried to rebase to main to get rid of the already merged PR, but this did not work.

My main concern is that it is still not clear to me how this is supposed to be used. This api does not make sense to me. How would a user know they would need to run this?

```python
structure = load_known_material('argyrodite')
vol.nearest_structure_reference(structure)
path.cartesian_path(vol)
path.fractional_path(vol)
path.path_over_structure(structure, vol)
plots.energy_along_path(path=path)
```

vs.

```python
plots.energy_along_path(path=path)
```

What is the difference?

stefsmeets commented on src/gemdat/path.py:
```diff
@@ -15,13 +19,63 @@ class Pathway:
+            raise ValueError('Voxel coordinates of the path are required.')
+        # Manually get the fractional coordinates of the sites, then use
+        # lattice.get_fractional_coords and make into tuple
+        self.frac_sites = [
```
```suggestion
        return [
```

SCiarella replied:
done

SCiarella commented on src/gemdat/path.py:
```diff
@@ -15,13 +19,63 @@ class Pathway:
+                fractional_coords)
+            self.cart_sites.append(tuple(cartesian_coords))
+
+    def fractional_path(self, vol: Volume):
```
done, thanks!

SCiarella commented:
Notice that I have not implemented the `@cache` decorator for `fractional_path`, `cartesian_path` and `nearest_structure_reference`.
The implementation is not trivial because our custom `Volume` and `Pathway` classes are not hashable.
